std/experimental/typecons.d

```diff
@@ -1074,3 +1074,307 @@ pure nothrow @system unittest
+ * subclass of `Exception` (which is the default).
+ *
+ * See_Also: https://www.youtube.com/watch?v=nVzgkepAg5Y
+ * See_Also: https://github.com/dlang/phobos/pull/6665
+ *
+ * TODO https://dlang.org/phobos/std_typecons.html#.apply
+ *
+ * TODO provide a member function `value()` or `expectedValue()` aswell?
+ *
+ * TODO I'm not convinced about the naming
+ * - `Expected`: instead call it something that tells us that it can be either expected or unexpected?
+ * - `Unexpected`: if so why shouldn't we have a similar value wrapper `Expected`?
+ *
+ * TODO we could get around the `Unexpected` wrapper logic by instead expressing
+ * construction in static constructor functions, say,:
+ * - static typeof(this) fromExpectedValue(T expectedValue)
+ * - static typeof(this) fromUnexpectedValue(E unexpectedValue)
+ *
+ * TODO swap
+ *
+ * TODO which functions should be `nothrow`?
+ *
+ * TODO later on: remove _ok when `_expectedValue` and ` _unexpectedValue` can store this state
+ * "collectively" for instance when both are pointers or classes (use trait
+ * `isAddress`)
+ *
+ * TODO ok for default construction to initialize
+ * - _expectedValue = T.init (zeros)
+ * - _ok = true (better to have _isError so default is zero bits here aswell?)
+ */
+struct Expected(T, E = Exception)
+if (!isInstanceOf!(Unexpected, T)) // an `Unexpected` cannot be `Expected` :)
+{
+    import std.algorithm.mutation : moveEmplace;
+
+    /// Construct from expected value `expectedValue.`
+    this(T expectedValue) @trusted
+    {
+        moveEmplace(expectedValue, _store.expectedValue);
+        _ok = true;
+    }
+    /// ditto
+    this(ref T expectedValue)
+    {
+        _store.expectedValue = expectedValue; // TODO do we need to care about `_store.expectedValue` init value??
+        _ok = true;
+    }
+
+    /// Construct from unexpected value `unexpectedValue.`
+    this(Unexpected!E unexpectedValue) @trusted
+    {
+        moveEmplace(unexpectedValue, _store.unexpectedValue);
+        _ok = false;
+    }
+    /// ditto
+    this(ref Unexpected!E unexpectedValue)
+    {
+        _store.unexpectedValue = unexpectedValue; // TODO do we need to care about `_store.unexpectedValue` init value??
+        _ok = false;
+    }
+
+    /// Assign from expected value `expectedValue.`
+    void opAssign()(auto ref T expectedValue) @trusted
+    {
+        // TODO is this ok?:
+        clear();
+        moveEmplace(expectedValue, _store.expectedValue);
+        _ok = true;
+    }
+
+    /// Assign from unexpected value `unexpectedValue.`
+    void opAssign(E unexpectedValue) @trusted
+    {
+        clear();
+        moveEmplace(unexpectedValue, _store.unexpectedValue);
+        _ok = false;
+    }
+
+    /// Clear (empty) contents.
+    private void clear() @trusted
+    {
+        release();
+        static if (isAddress!T)
+        {
+            _store.expectedValue = null;
+        }
+    }
+
+    /// Release any memory used to store contents.
```

Comment:
Misleading comment. This function calls `destroy` on the contents, which may free memory, or may free other resources, depending on what type is being stored. The memory used by the `Expected` instance itself is not released.